apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -1568,12 +1592,24 @@ maybe_send_element_safe(State, El) ->
         _ -> error
     end.
 
-send_element(#state{server = Server, sockmod = SockMod} = StateData, El)
-  when StateData#state.xml_socket ->
-    ejabberd_hooks:run(xmpp_send_element, Server, [Server, El]),
+%% @doc This is the termination point - from here stanza is sent to the user
+%% We sent the original stanza ('element') unless there is a different thing
+% keyed 'to_send'
+-spec send_element(mongoose_acc:t() | state(), state() | xmlel()) -> mongoose_acc:t() | ok.
+send_element(#state{} = StateData, #xmlel{} = El) ->
+    ?DEPRECATED,
+    send_element(mongoose_acc:from_element(El), StateData),
+    ok;
+send_element(Acc, #state{server = Server} = StateData) ->
+    Acc1 = ejabberd_hooks:run_fold(xmpp_send_element, Server, Acc, [Server]),
```

Comment:
> or go the other way and treat an acc entering the recipient's c2s as something brand new

Which actually is the case for S2S messages. From other server we get only stanza without any other data.
I think there is still value in passing the Acc between different C2S process. The data inside the Acc should be limited though. Right now I think none of the data we pass is useful for the recipient, please correct me and show me what exactly is beneficial for the receiver. I'd like to be wrong about it.
The value I'm talking about is purely for debugging and performance analyzation. At this moment I think the data passed between c2s process should be:
* timestamp when the Acc was created in the sender's c2s
* and internal ID - for debugging purposes so that we can track how the Acc travels inside MongooseIM